va/varenderer.cpp

```diff
@@ -0,0 +1,680 @@
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <dlfcn.h>
+#include <errno.h>
+#include "cros_gralloc_handle.h"
+#include "vautils.h"
+#include <pthread.h>
+#include "utils/autolock.h"
+#include <string.h>
+// #include "drmhwcomposer.h"
+#include "hwc2_device/HwcLayer.h"
+namespace android {
+#define ANDROID_DISPLAY_HANDLE 0x18C34078
+#define CHECK_SYMBOL(func) { if (!func) printf("func %s not found\n", #func); return VA_STATUS_ERROR_UNKNOWN; }
+#define DEVICE_NAME "/dev/dri/renderD128"
+#ifdef LOG_TAG
+#undef LOG_TAG
+#define LOG_TAG "hwc-varender"
+#endif
+VARenderer::~VARenderer() {
+  DestroyContext();
+  if (va_display_) {
+    vaTerminate(va_display_);
+  }
+  DRMHwcNativeHandle temp_handle;
+  if(native_handles.size() == NATIVE_BUFFER_VECTOR_SIZE){
+    for( int32_t i=0; i<NATIVE_BUFFER_VECTOR_SIZE; i++ ) {
+      temp_handle = native_handles.at(i);
+      buffer_handler_->ReleaseBuffer(temp_handle);
+      buffer_handler_->DestroyHandle(temp_handle);
+    }
+    native_handles.clear();
+  }
+  if(native_rotation_handles.size() == NATIVE_BUFFER_VECTOR_SIZE){
+    for( int32_t i=0; i<NATIVE_BUFFER_VECTOR_SIZE; i++ ) {
+      temp_handle = native_rotation_handles.at(i);
+      buffer_handler_->ReleaseBuffer(temp_handle);
+      buffer_handler_->DestroyHandle(temp_handle);
+    }
+    native_rotation_handles.clear();
+  }
+  if(native_active_handles.size() == NATIVE_BUFFER_VECTOR_SIZE){
+    for( int32_t i=0; i<NATIVE_BUFFER_VECTOR_SIZE; i++ ) {
+      temp_handle = native_active_handles.at(i);
+      buffer_handler_->ReleaseBuffer(temp_handle);
+      buffer_handler_->DestroyHandle(temp_handle);
+    }
+    native_active_handles.clear();
+  }
+  ReleaseCache();
+}
+
+bool VARenderer::Init(uint32_t fd) {
+  unsigned int native_display = ANDROID_DISPLAY_HANDLE;
+  buffer_handler_.reset(NativeBufferHandler::CreateInstance(fd));
+  VAStatus ret = VA_STATUS_SUCCESS;
+  va_display_ = vaGetDisplay(&native_display);
+  if (!va_display_) {
+    ALOGE("vaGetDisplay failed\n");
+    return false;
+  }
+  ret = pthread_mutex_init(&lock_, NULL);
+  if (ret)
+    ALOGE("Failed to initialize the mutex lock %d\n", ret);
+  int major, minor;
+  ret = vaInitialize(va_display_, &major, &minor);
+  if (ret == VA_STATUS_SUCCESS) {
+    AllocateCache(DEFAULT_LAYER_NUM);
+  }
+  return ret == VA_STATUS_SUCCESS ? true : false;
+}
+
+void VARenderer::ReleaseCache() {
+  if (va_buffer_id_) {
+    free(va_buffer_id_);
+    va_buffer_id_ = nullptr;
+  }
+  if (surface_in_) {
+    free(surface_in_);
+    surface_in_ = nullptr;
+  }
+  if (surface_region_) {
+    free(surface_region_);
+    surface_region_ = nullptr;
+  }
+  if (output_region_) {
+    free(output_region_);
+    output_region_ = nullptr;
+  }
+}
+
+bool VARenderer::NeedResizeCache(uint32_t layer_num) {
+  return layer_num > layer_capacity_;
+}
+
+bool VARenderer::AllocateCache(uint32_t layer_capacity) {
+  surface_in_ = (VASurfaceID*)malloc(sizeof(VASurfaceID) * layer_capacity);
```

Comment:
Why not use `std::vector` to manage these resources as we are writing C++ here.

Reply:
porting from previous, now we need poc and do some test, next will refine